doc/ERROR_CONCEPT.md

```diff
@@ -0,0 +1,250 @@
+* Awareness of developers: All devs have to know all categories and choose the correct one. Devs should also be prohibited to throw their error
+into the "uncategorized" category due to lazyness.
+
+Suggestion:
+This line is by far the most discussable one. I thought about categorizations such as
+* Permission/ Resource/ Parse/ External(C++, Ruby, Java)/.../ Miscellaneous:
+Most errors are resource errors (could not stat, no connection, could not close/open/find etc.) This again would not give users
+a particularly useful error indication since he/she has to lookup in the reason anyway. Even subdividing the `Resource` category
+would lead to the same problem. No matter how I look at it, the user
+will have to read the concrete reason in most cases which makes the Description line not that valuable any longer.
+* temporary/permanent/conflict/validation:
+This seems to be a better approach because it includes a possible solution for the user. How to categorize certain errors though
+seems impossible because there can be multiple reasons where to categorize an error. Take `Could not open file` as an example:
+if it is on a NFS and it is offline, *temporary* would be correct. If the file has too restrictive permissions, *permanent* would be more
+appropriate since the admin has to fix it. If the file is currently in use and semlocked by the system, *conflict* would be appropriate.
+* core kdb error/plugin error:
+Not much value for the user
+* differentiation to core features of elektra: Specification/ Notification/ KDB Core/ Bindings/ ...
+An error is again difficult to put into one bucket since some errors might be affected by multiple features (could not stat file for example
+which is used by multiple features).
+This approach would lead to the same problem as we have now (many errors, difficult to maintain)
+
+Personally I do not see how the benefit of a categorization outweighs the drawbacks because in by far the most cases users will have to look
+into the reason in more detail anyway. It should be enough for users to deal with the error just by the reason.
+
+As a result I would suggest to remove the description line.
+
+### Line 2: Error code
+
+Reasons to keep it:
+* Probably could be used to categorize errors such it is done in the http protocol
+* Good to use for shellrecorder scripts to test for errors
+* Possibility for automation (eg. if a certain error number occurred you could automatically trigger certain actions)
+
+Reasons against it:
+* Extra line
+
+Suggestion:
+Giving an error a number like it is done in C programs as return code seems to be outdated for elektra, especially when
+the reason is given anyway. Personally I never read the error number because it has not much value.
+The *reasons to keep* will also be very limited when the number of errors are significantly reduced.
+I would also suggest to remove it.
```

Comment:
It is needed for automated checks for specific errors (especially unit and shell recorder tests).

What we could do is to not use numbers but some ID, like the macro name or alternatively something generated (which is guaranteed to be unique). This would avoid conflicts when adding errors and make it easier for external specifications.

Downside: string comparison is needed to check for which error it is. Maybe that is not really a problem.

Reply:
I am confused by your message:
Is it absolutely needed or not? As of now the unit tests and shell recorder tests test for a very specific error number. If we reduce the number of errors (which is the main purpose here) we would have to change unit tests anyway.

Does it make such a difference when comparing for a String or Number? If the error message changes, the test too but do error messages get changes that much?

If we could get rid of the `specification` file we can let developers handle errors much more easier.

Reply:
In my opinion an error code is **absolutely essential**. There has to be a way to infallibly check which error occurred, because some errors may be fixed automatically without user input. The high-level API for example retries kdbSet, if a conflict (error code 30 I think) occurred, with the user not even knowing about it.

Think of it this way:
Most modern languages (e.g. C++ and Java) support exceptions. Certain exceptions (e.g. a timeout for an HTTP request) can be caught and fixed in some way (retry HTTP request with exponential backoff). Others might be totally ignored/suppressed or wrapped in a more general exception. For this reason it is good practice to use different exception types. You could of course always use the same type and compare the error message string, but this is not reliable. A single character changed in the error message breaks the comparison. Not even a typo could be fixed without breaking backwards compatibility. And that doesn't even consider the very common case of dynamic information in an error message. If the error message contains a filename that depends on user input (e.g. mountpoint), your only chance is to partially compare the string. This partial comparison is even more error prone, because it might match other errors as well.

C doesn't have exceptions, so other values (integers in most cases) are used to uniquely identify one type of error. Too guarantee the error code is unique, a centralised specification is the easiest way.

However, I think this specification can be vastly improved:
- Don't specify error codes explicitly. Simply start with 1 for the first error and go from there. This eliminates merge conflicts, as long as we only use macros (or other named symbols) in code.
- Don't use macros. Use `extern const int`. This way the actual error codes we use are hidden from external code and are determined by the version of Elektra used. (@markus2330 please correct me if this is wrong)
- If we also want to guarantee that the actual error codes never change, we could fixate them on each release. Either explicitly, or by using a simple test case that `git pull`s the last release and compares error codes.
- Only put the names for the `extern const int`s and a short description/comment what this error code should be used for into the specification. This allows the error message to be fully customised and an error code would be more or less equivalent to an exception type in Java or C++. The format would basically be:
  ```
  name: FILE
  description: an error related to a file, e.g. missing permissions or missing file
  ```
  From that we would generate `kdberror.h`:
  ```c
  extern const int ELEKTRA_ERROR_FILE;
  ```
  And a source file:
  ```
  #include <kdberror.h>
  
  /**
   * an error related to a file, e.g. missing permissions or missing file
   */
  const int ELEKTRA_ERROR_FILE = 1;
  ```

Reply:
What you suggest though is basically what we have now, hundreds of error codes in a centralized file to distinguish them among each other. Many error codes even share the same text.

So in your opinion this task is just to squash some errors together? This would lead to the same problems again over the long run which we have now. You want to know if there has occurred a very specific error so you can actually do counter measures. My task though was to reduce the high granularity of the error messages.

Where should this be written? 
```
name: FILE
description: an error related to a file, e.g. missing permissions or missing file
```
In the specification file? Or is the kdberror.h the new source of error codes to be checked for rather then the specification.

Reply:
> What you suggest though is basically what we have now

No, I agree with you on all parts of your concept except error codes. Like Markus said above, we don't have to use integers, but we do need some sort of unique ID.

> hundreds of error codes

We don't need hundreds of codes, but we need unique codes for some errors. I personally would group the rest thematically: file errors, broken specification, invalid configuration (i.e. not in accordance to the specification), error in binding (e.g. Java exception) etc.

Looking at the current specification I see only three things that should definitely have separate error codes:
- `CONFLICT` (30): can sometimes be resolved by retrying
- `TIMEOUT` (195): similarly might be resolved by retrying (may be with bigger timeout)
- `NETWORK` (205): An application could suppress this error, wait until a connection is available again and retry then. Of course some sort of message may still be produced to inform the user.

All of these basically come down to the same behaviour: "auto-retry". Especially, if an application is running in the background it is very nice, if the application doesn't require user input for things that might fix themselves. Auto-retry is simply not possible, if you don't have an infallible way of identifying an error.

> in a centralized file to distinguish them among each other.

I really don't care whether the file is centralized, or whether there is a file at all. If you can find a different solution to ensure unique error codes, I'm okay with that too. We could even use code 1 for conflict, 2 for timeout, 3 for network, ... (any future "auto-retry" codes) and -1 for anything else. Basically defining positive error codes are auto-retry-able, negative ones are not. I really don't care, as long as the auto-retry-able errors are infallibly distinguishable from the rest and from each other.

> Where should this be written?

Probably it would be better to just write `kdberror.h` and the corresponding source file by hand. It should only be a few lines anyway.

> This would lead to the same problems again over the long run which we have now.

As long as we make sure that each new error code introduced in a PR is very well justified, I don't think this will happen.

Reply:
@kodebach wrote:
> However, I think this specification can be vastly improved:

I fully agree! These may seem as little low-level problems but they are very annoying in the day-to-day programming.

>  "auto-retry".

I think we should limit "auto-retry" to conflicts (where it is clear what to do). Otherwise auto-retries are better handled by plugins, which have domain specific knowledge (contrary to the applications).

Nevertheless, it is not only useful to distinguish "auto-retry" and "abort" but also between "resource" and "validation" errors: in one case changing the values may help, in the other not.

> Probably it would be better to just write kdberror.h and the corresponding source file by hand. It should only be a few lines anyway.

I do not agree here. I do not see how it could be reduced to "a few lines". Furthermore, the specification will provide immense benefits once we properly support other compiled languages (like C++, Java, Rust, Go, ...). Ideally @Piankero could write a translator to C++ and Java Exception so that we know that the specification has enough information to generate exceptions.

Reply:
> I think we should limit "auto-retry" to conflicts (where it is clear what to do). Otherwise auto-retries are better handled by plugins, which have domain specific knowledge (contrary to the applications).

Just to clarify: "auto-retry" was just an example of why it is sometimes useful for an application to identify the specific error that occurred. Whether or not actually doing an automatic retry makes sense in a certain application and situation is a totally different thing.

> I do not see how it could be reduced to "a few lines".

It would only be a few lines, if we only defined the error codes (plus a code comment for description) in the central file. As soon as we put error codes into categories or add other properties, the specification makes sense again.

> the specification will provide immense benefits once we properly support other compiled languages

I am not sure automatically generating code for other languages is a good solution. For one Rust and Go don't have exceptions and handle recoverable errors more like functional languages (by returning multiple values, a value and/or an error). But even if the language in question has exceptions, auto generating them is not wise. Sometimes using a standard library exception would be preferable. Sometime you might want to subclass another exception type, so you can catch many exceptions or just a special one. And all that aside, to "properly support" a language the binding has to be hand written anyway, otherwise it will always feel clunky and foreign. So you might as well write the exception classes by hand as well (maybe starting from a one-time auto-generated version).

Reply:
> "properly support" a language the binding has to be hand written anyway,

This is not the experience we have had so far. All bindings except of Java are auto-generated, and the Java binding is definitely not the best one.

> So you might as well write the exception classes by hand as well

Of course you can do that but then you risk inconsistencies.

Reply:
Ok, after some longer brainstorming I came to the conclusion that we first should agree on the categories to put all current errors into. For users, these categories do not really have a special meaning because they just care about the concrete error reason anyway. That's why I suggested to remove it. The only reason is keep it is to **differ for programmatic reasons**.

Now the question is what exactly are those categories. For me, every category should dictate another solution"concept" such as, retry, unfixable, try something different, ...
My first proposal is the following:

- **Validation**: Try a different value
- **Temporary**: Retry it
- **Permanent**: Fixing is impossible or a very special case

Is there any other big category anybody knows of which would require a different approach? (@markus2330 , @kodebach , @sanssecours? Since you have the most experience I would value your input very much)

I did not see a specific difference in Temporary and Conflicts as Markus suggested in #1473 because they both are basically retry approaches but one is more "harmful".

The problem here is the granularity as sometimes even small cases such as "could not open file" (which currently would be under permanent errors) can be fixed but then would require String parsing to check for this specific error. If it is really the case that a developer wants to know a very specific error, I think it is acceptable to force String parsing. Also some errors with the same reason could fall under "temporary" (timeout) or permanent ("network driver failure"). The developer has to know in this case which category he has to choose because he has the concrete domain knowledge.

I also thought about auto generated error codes but beside the question if it is even implementable in C, it comes with other drawbacks as developers would then need to check for too many errors (eg. think of specification errors: is it an enum/path/boolean/etc. error.). To me it looks like the error concept is a fight in the level of granularity with each level coming with its own advantages/drawbacks. For me the current approach covers the most central cases with openness to new categories, should they be needed at a later point in time.

I also worked out some sort of migration of the current errors: https://docs.google.com/spreadsheets/d/1-vXNZ7pN9wlMFByIMLbFt_HieyJpop0UyksbgAvmGK4/edit?usp=sharing

I know you are all no fan of google and excel sheets but since this is still very hybrid it would take considerably longer to do it in markdown every time something changes. Once the categories + mappings are clear I will incorporate it into the README. Feel free to comment or change it as I am also not 100% sure if it is a temporary or permanent error in many cases.

Reply:
Thank you for the write-up! I think you can rework the whole error concept now. Maybe start with the goals, which should be very clear now. Furthermore it would be nice if you directly rework the error specification, the spread-sheet already looks good. I think we can split the categories a bit more:

- Validation into syntactic and semantic (and maybe even others)
- Temporary into conflicts and timeouts (watch out: only because "temporary" is in the description does not mean it is a temporary problem: 161 is a resource/permission error)
- Permanent into Resource, Parsing, Installation errors (and maybe even others)

Maybe we want sub-categories, so that applications which e.g. do not care if a validation is syntactic or semantic can write error handling code easier.

I think there is little use of discussing this even more. Simply do it! (I would start with categorizing the errors directly in the error specification. It is a harmless thing and can be merged easily as it will not break anything.)

Reply:
I forgot a very important class of errors: Assertion failures! I.e. logical errors that should never happen and are bugs within Elektra.

Reply:
I think I finally know why we disagree so much. Who is a user in your mind? 
 - A user of Elektra, i.e. a developer using Elektra in their application.

or
 - A user of an application using Elektra (e.g. the `kdb` tool).

The first kind (let's call them "application developers") will probably find specific error codes useful. The more specific an error code is, the easier you can find documentation on how to fix it. While a good description might be enough on its own, an error code always helps, if you need look at the documentation.

The second kind of users (let's call them "end users") won't find that useful. They just want a single line that tells them want they need to do "to make it work". This includes users of `kdb`, so I'm going to use it as an example.

You rightfully pointed out that `kdb`s error messages are quite bad, mainly because of the huge amount of unnecessary (and partly redundant) information. But this has nothing to do with how Elektra itself reports errors to applications. Even if we kept everything the same, `kdb` could just print

```
Error 121: validation failed
Validation of key "<key>" with string "<value>" failed.
```

instead of the 9 lines you mentioned in your example. (`-v` is used, in which case the second *verbose* version makes sense)

---

So I think the problem is twofold and we need two separate decisions here. The questions are:
1. How should Elektra (the library) report errors to applications?
2. How should `kdb` display these errors?

The second question is easy to answer: `kdb` should print one or two lines of error description. `kdb -v` should print **all** the available information, so that diagnosing errors becomes easier.

The first question is not so easy. The important thing here is to get rid of all the duplicate or very similar errors (e.g. the GPG errors), while maintaining enough granularity, so that error codes are useful for testing, diagnosing and actually handling errors. The categories are a good start, but IMO they can't replace actual error codes. We don't need an separate code for each description (in fact that is exactly what we shouldn't do), but we need some distinction.

Another good step to get rid of unnecessary error codes is IMO to not have any kind of specification for warnings. The difference between a warning and an error is that a warning can (more or less) safely be ignored, so I don't see any reason to specify warnings. If we just use `elektraAddWarning(Key * key, const char * format, ...)`, we would get rid of 84 error codes immediately. NOTE: warnings also shouldn't translate into exceptions or similar in bindings no need for specification there either

Reply:
> I think we can split the categories a bit more

> Maybe we want sub-categories

More categories plus sub-categories, might be an alternative to error codes, but I'm not quite convinced it would be a good fit for testing purposes.

> Assertion failures! I.e. logical errors that should never happen and are bugs within Elektra.

Should those even be errors? Shouldn't we just exit if an assertion fails? If this "should never happen," crashing the application shouldn't really be a problem, should it?

Reply:
I fully agree we should distinguish of which error information we emit and which parts of this information is shown to whom.

> Should those even be errors? Shouldn't we just exit if an assertion fails?

The application should exit but a well behaved library should not exit. (The infallible APIs as exception.) So I think we need (assertion) errors for the situations where admins installed Elektra incorrectly or the KDB is completely broken (random stuff in system/elektra).